### ISSUE TITLE
Nested navigation with redux

### ISSUE THREAD

**Issue by MyTotoro** [react-community/react-navigation#1041]
Hi, I've gotten redux integrated and have the parent level navigator (stack) pretty much figured out. The issue I'm trying to fix is now to integrate/use the nested navigator (drawer)

```
MyStackNavigator
-Splash
-Login
-MainDrawerNavigator

MainDrawerNavigator
-Home
-Settings
-etcscreen
```

I've used reactotron and my state tree looks like this:
![image](https://cloud.githubusercontent.com/assets/3621921/25032896/2f8dfd0e-20a6-11e7-8749-acd034972654.png)

My question is how to handle

1. programmatically opening the drawer
2. navigating to another screen (settings for example)
3. reset navigation stack completely and go back up to Splash screen in parent navigator

I've tried increasing the index to 1 so that DrawerOpen is selected but that didn't seem to accomplish anything:

```
const newState = Object.assign({}, state)
newState.routes[0].index = 1
//return MainNavigation.router.getStateForAction(navigateAction, state)
return newState
```

I looked around on the issues and found that one user suggested using:

```
this.props.dispatch({
  type: 'Navigate',
  routeName: 'Home',
  action: {
    type: 'Navigate',
    routeName: 'Main',
  },
```

I've tried integrating this and didn't get anywhere

**Comment by samridhgupta** [comment]
+1

**Comment by DongShi** [comment]
+1

**Comment by lorenzsell** [comment]
+1

**Comment by byk04712** [comment]
+1

**Comment by asmurianwar** [comment]
+1

**Comment by zacholas** [comment]
+1

**Comment by dinhha** [comment]
+1

**Comment by X-Tender** [comment]
Did someone have an Example with Redux + Nested Navigations?

**Comment by RUIFERNANDE5** [comment]
+1
I need to be able to reset a structure like the following:

```
Auth:
  - Login
  - Register
Main
  - Dashboard
```
I can only navigate from Login to Main/Dashboard with dispaching navigate action.
How can I achieve this with reset?

**Comment by cooperka** [comment]
Related: https://github.com/react-community/react-navigation/issues/937. There don't seem to be docs for any of these use cases :(

State also seems to be really weird when nesting routers. I wonder if there's just a bug.

For example, a [DrawerNav nested inside a StackNav](https://github.com/cribspot/react-navigation/commit/51e607cc910820e2e0f4683e0f104c887c3ccc41) gives this simple initial state (using `MyNav.router.getStateForAction(MyNav.router.getActionForPathAndParams('RouteName'))`):

```js
{
  index: 0,
  routes: [
    {
      routeName: StackA,
      key: Init
    }
  ]
}
```

but a [StackNav nested inside a DrawerNav](https://github.com/cribspot/react-navigation/commit/d9672213f8365ab9601548c0a76da507d4e3bf0d?diff=split) gives this monolith of an initial state:

```js
{
  routes: [
    {
      routes: [
        {
          index: 2,
          routes: [
            {
              routeName: StackB,
              key: Init
            },
            {
              key: id-1492717322143-0,
              routeName: StackB
            },
            {
              key: id-1492717322143-1,
              routeName: StackB
            }
          ],
          key: DrawerA,
          routeName: DrawerA
        }
      ],
      index: 0,
      key: DrawerClose,
      routeName: DrawerClose
    },
    {
      key: DrawerOpen,
      routeName: DrawerOpen
    }
  ],
  index: 0
}
```

I can't imagine why they're so different. Both seem wrong to me; the first seems to be missing information and the second seems to be duplicating things.

**Comment by MyTotoro** [comment]
@cooperka actually I think your first state tree is fine 

A drawer nav is very flat (screen1, screen2, drawerState1, drawerState2). The first picture that I posted was a drawer navigator nested inside a stack navigator

I'll be digging around this weekend to figure this out - I think the crux of the problem is that the ACTION being fired is not behind handled properly:

1. people who are implementing the bare minimal getStateForAction are doing it only for the top level router (toplevel.router.getStateForAction...) - where the base case is returning the unchanged state hence the usual 'nothing happens'

2. the reducer is not properly associated with the fired ACTION - in my example, the default type fired is 'Navigation/NAVIGATE', I'm using redux-sauce and I can't even type forward slashes in the action definitions

**Comment by lorenzsell** [comment]
I have a similar set up to @MyTotoro with login in a stack nav that has a nested drawer nav.  After the user logs in,  I use the reset action to reset the index to 0 on the drawer nav:

``
const loginSuccessAction = NavigationActions.reset({
        index: 0,
        actions: [
          NavigationActions.navigate({ key: 'Main', routeName: 'Main' })
        ]
      });
      return InitialNavigator.router.getStateForAction(loginSuccessAction, state);
``

On the login page, I check to see if the user is already logged in, and, if they are, I also call the reset action.

Not sure if this is the right way to do it, but it does work seamlessly now.

**Comment by MyTotoro** [comment]
@lorenzsell Does that allow you to navigate the drawer properly?

Are both the stack and drawer navigators handled by the same reducer?

**Comment by lorenzsell** [comment]
Yes, I have one navigation reducer that handles all of my navigators.  My drawer uses a custom content component.

Here is my set up and what my reducer looks like.  This took me a long time to get working right, so hopefully it can be helpful to you.

+ StackNavigator
  + Login
  + CreateItem
  + Main (DrawerNavigator)
    +PodListNav (StackNavigator)
      +PodList
      +PodNavigator (DrawerNavigator)


```
const INITIAL_STATE = InitialNavigator.router.getStateForAction(InitialNavigator.router.getActionForPathAndParams('Login'));

export default (state = INITIAL_STATE, action) => {
  switch (action.type) {
    case USER_LOGGED_IN:
      return InitialNavigator.router.getStateForAction(
        NavigationActions.navigate({ routeName: 'Main',
                                     action: NavigationActions.navigate({
                                       routeName: 'PodListNav',
                                     }),
                                   }), state);
    case LOGIN_USER_SUCCESS:
      const loginSuccessAction = NavigationActions.reset({
        index: 0,
        actions: [
          NavigationActions.navigate({ key: 'Main', routeName: 'Main' })
        ]
      });
      return InitialNavigator.router.getStateForAction(loginSuccessAction, state);
    case LOGOUT_USER:
      const logoutAction = NavigationActions.reset({
        index: 0,
        actions: [
          NavigationActions.navigate({ routeName: 'Login' })
        ]
      });
      return InitialNavigator.router.getStateForAction(logoutAction, INITIAL_STATE);
    case SELECT_POD:
      return InitialNavigator.router.getStateForAction(NavigationActions.navigate({ routeName: 'PodNavigator' }), state);
    case SELECT_SPACE: // This is a drawer action
      return InitialNavigator.router.getStateForAction(NavigationActions.navigate({ routeName: 'PodListNav' }), state);
    case CREATE_ITEM:
      return InitialNavigator.router.getStateForAction(NavigationActions.navigate({ routeName: 'CreateItem' }), state);
    default:
      return InitialNavigator.router.getStateForAction(action, state);
  }

```

**Comment by RWOverdijk** [comment]
this is showing some weird behavior here. I have tried using a reducer for each router, at first. Then I tried having a single reducer. The behavior is identical (turns out I didn't need multiple reducers).

I have a drawer navigator, with three nested stack navigators. If, in one of the nested stack navigators, I go to a page inside of that stack, and press back, I get sent to the root page of the drawer.

My guess, is that it has something to do with how it manages the stack when inside of a drawer: there appears to be none. No index, either. Something here is off, or poorly implemented by me. I'd love some more clarity on this, too.

TL;DR;
👍

**Comment by MyTotoro** [comment]
@RWOverdijk 

From the docs:
```
Keep in mind that when a navigator is given a navigation prop, it relinquishes control of its internal state. That means you are now responsible for persisting its state, handling any deep linking, integrating the back button, etc.

Navigation state is automatically passed down from one navigator to another when you nest them. 
```

Your implementation seems to be _correct_ however not _predictable_.

@lorenzsell your `InitialNavigator` is the StackNavigator?

```
Main (DrawerNavigator)
+PodListNav (StackNavigator)
+PodList
+PodNavigator (DrawerNavigator)
```

So you have a Drawer with 3 child screens - does having a custom content component affect anything?

**Comment by braco** [comment]
@lorenzsell thanks for that code. Isn't it an antipattern to have control functionality (navigation commands) in the reducer? This would normally go into a side effect library like sagas, observables, thunks, etc.

**Comment by MyTotoro** [comment]
@braco Not in particular - if your app is simple enough that you don't need it - you shouldn't, don't integrate it into redux for the sake of doing so. But if you are using sagas to `takeLatest` a specific type of navigation action and doing some updating, it can make reasoning and implementation much easier to work with

@lorenzsell I've implemented a working version but I'm having trouble resetting - how do you reset the parent stack? The child router always complains that the key I'm trying to reset to doesn't exist

For those wondering how to implement nested routers. This logic should work for any variant of nested routers. I've connected my app to redux as I'll be hooking navigation actions later on with sagas. I use reduxsauce and seamless-immutable because that came with the template, but try not to focus on the syntax

My structure looks like:

```
Main App (Stack):
 - Login
 - Main (Drawer)  

Main (Drawer)    
 - Home
 - Settings (Stack)

Settings (Stack)  
- Settings1
- Settings2

```

Define your parent router and connect it to redux:
```
render () {
    return (
      <StackNavigationRouter
        navigation={
          addNavigationHelpers({
            dispatch: this.props.dispatch,
            state: this.props.navigation
          })
```
Now state.navigation is the state management that the reducer will modify

The StackNavigationRouter itself looks like:
```
import DrawerNavigationRouter from '../.....'

const routeConfiguration = {
  Splash: {
    screen: SplashScreen
  },
  Login: {
    screen: LoginScreen
  },
  Main: {
    screen: DrawerNavigationRouter
  }
}
...
const StackNavigationRouter = StackNavigator(
  routeConfiguration,
  customOptions
)

export default StackNavigationRouter
```
Yes, `DrawerNavigationRouter` is a drawer router

```
const DrawerNavigationRouter = DrawerNavigator(
  {
    Home: {
      screen: MainContainer
    },
    Settings: {
      screen: SettingsNavigationRouter //this is a stack router
    }
  },
  {
    drawerWidth: width,
    ...  
}
)

export default DrawerNavigationRouter
```

My reducer listeners look like this (mainly because of sagas)
```
export default () => {
  /* ------------- Assemble The Reducers ------------- */
  const rootReducer = combineReducers({
    login: require('./LoginRedux').reducer,
    navigation: require('./NavigationRedux').reducer,
    ...

  })

  return configureStore(rootReducer, rootSaga)
}
```
If you are familiar with integrating with Redux - you are essentially using the reducer in `NavigationRedux` to handle the navigation state

`NavigationRedux` is created with the help of redux sauce
```
const { Types, Creators } = createActions({
  navigateMain: null,
  navigateLogin: null,
...

export const INITIAL_STATE = Immutable({
  index: 0,
  routes: [{
    routeName: 'Splash',
    key: 'init'
  }]
})

export const navigate = (state = INITIAL_STATE, action) => {
  console.tron.log('Inside NavigationRedux navigate function')
  var navigationAction = {}
  switch (action.type) {
    case Types.OPEN_DRAWER:
      StackNavigationRouter.router.getStateForAction(NavigationActions.navigate({ routeName: 'DrawerOpen' }), Immutable.asMutable(state, {deep: true})))})
      navigationAction = NavigationActions.navigate({
        routeName: 'DrawerOpen'
      })
StackNavigationRouter.router.getStateForAction(navigationAction, Immutable.asMutable(state, {deep: true})))})

      return Immutable.merge(state, StackNavigationRouter.router.getStateForAction(navigationAction, Immutable.asMutable(state, {deep: true})))

    case Types.NAVIGATE_LOGIN:
        return Immutable.merge(state, StackNavigationRouter.router.getStateForAction(NavigationActions.navigate({ routeName: 'Login' }), Immutable.asMutable(state, {deep: true})))

...

export const reducer = createReducer(INITIAL_STATE, {
  [Types.NAVIGATE]: navigate,
  [Types.NAVIGATE_MAIN]: navigate,
  [Types.NAVIGATE_LOGIN]: navigate,
  'Navigation/NAVIGATE': navigate  //yes, if you pull the drawer open, the drawer itself will fire this type, you need to catch it with your reducer to handle it properly / maintain sync with your state tree in redux
```

Any more nested routers follow the same pattern, one thing to point out is that the reset actions will error out (at least with the tests that I've done so far). What I mean by that is

```
Main App (Stack):
 - Login (Trying to go here by resetting and navigating to 'Login')
 - Main (Drawer)  

Main (Drawer)    
 - Home
 - Settings (Stack)

Settings (Stack)  
- Settings1 (You are here)
- Settings2

```

I have a BackHandler exiting the app at Main/Login so that you can never 'back-press' into a logged in state so it works for me

**Comment by AlmogRnD** [comment]
Has anyone got this working I'm having lot's of issues with this, just tried to trigger the open drawer method from on press on bars icon? Here is my code 

My index.js 
```
// App state
const AppWithNavigationState = connect( state => ({ nav: state.nav})) (({ dispatch, nav }) => (
    <AppNavigator navigation = { addNavigationHelpers ({ dispatch, state: nav })} />
));


class App extends React.Component {
    store = createStore(AppReducer, undefined, autoRehydrate());

    componentDidMount() {
        persistStore(this.store, { storage: AsyncStorage });
    }

    render() {
        return (
            <Provider store={this.store}>
                <AppWithNavigationState />
            </Provider>
        );
    }
}
AppRegistry.registerComponent('App', () => App);
export default App;


const initialNavState = {
    index: 1,
    routes: [
        { key: 'Login', routeName: 'Login' },
        { key: 'ResetPassword', routeName: 'ResetPassword' },
        //{ key: 'Main', routeName: 'Main' },
    ],
};
```

app reducer
```
const initialAuthState = { isLoggedIn: false };
const AppReducer = combineReducers({
    nav: (state = initialNavState, action) => {
        switch(action.type) {
            case 'Login':
                return AppNavigator.router.getStateForAction(NavigationActions.back(), state);
            case 'Logout':
                return AppNavigator.router.getStateForAction(NavigationActions.navigate({ routeName: 'Login' }), state);
            case 'ResetPassword':
                return AppNavigator.router.getStateForAction(NavigationActions.navigate({ routeName: 'ResetPassword' }), state);
            case 'Main':
                return AppNavigator.router.getStateForAction(NavigationActions.navigate({ routeName: 'Main' }), state);
            default:
              return AppNavigator.router.getStateForAction(action, state);
        }
    }
});
export default AppReducer;

```
```
router
// Main drawer icon
const DrawerIcon = ({navigation}) => {
    return(
    <View>
        <TouchableOpacity onPress={() => {this.props.navigation.navigate('DrawerOpen'); } }>
            <Icon name="bars" style={{padding: 10, marginLeft:10}} size={20} color="black" type={"font-awesome"}/>
        </TouchableOpacity>
    </View>
    );
};

// Drawer Navigation
export const MainDrawerNavigator = DrawerNavigator({
    Home: {
        screen: Home,
        navigationOptions: {
            drawer:{
                label: "Home",
            }
        },
    },
    Logout: {
        screen: Logout,
        navigationOptions: {
            drawer:{
                label: "Logout",
            }
        },
    }
});

// Native tab navigation
export const TabsNavigation = TabNavigator({
    Home: {
        screen: Home,
        navigationOptions: ({ navigation }) => ({
            tabBarLabel: 'Home',
            title: "Home",
            headerLeft : <DrawerIcon navigation={navigation} />,
            tabBarIcon: ({ tintColor }) => <Icon name="home" size={28} color={tintColor} type={"font-awesome"} />
        }),
    },
    Photos: {
        screen: Photos,
        navigationOptions: {
            tabBarLabel: 'Photos',
            title: "Photos",
            tabBarIcon: ({ tintColor }) => <Icon name="camera" size={28} color={tintColor} type={"font-awesome"} />
        },
    },
    SiteDiary: {
        screen: SiteDiary,
        navigationOptions: {
            tabBarLabel: 'Site Diary',
            title: "Site Diary",
            tabBarIcon: ({ tintColor }) => <Icon name="cog" size={28} color={tintColor} type={"font-awesome"} />
        },
    },
});

// Main navigation StackNavigator
export const MainScreenNavigator = StackNavigator({
    Home: {
        screen: TabsNavigation
    },
    Drawer: {
        screen: MainDrawerNavigator
    }
});

// Main public nav
export const AppNavigator = StackNavigator({
    Login: {
        screen: Login,
        navigationOptions: {
            title: "Login",
            headerLeft: null,
        },
    },
    ResetPassword: {
        screen: ResetPassword,
        navigationOptions: {
            title: "Reset Password",
            headerLeft: null
        },
    },
    Main: {
        screen: MainScreenNavigator,
        navigationOptions: {
            header:null,
            gesturesEnabled: false,
        }
    }
});

```

**Comment by kelset** [comment]
Pinging OP @MyTotoro since this issue is quite old and related to an old version of the lib. If no answer I'll close in 7 days.

Moreover, we are currently working in adding more complex examples and in the meantime we added a new section in the README with links to complex examples.

**Comment by julesmoretti** [comment]
@kelset - thank you that would be great because its a real struggle and most examples out there are no longer functional or are missing documentation.

**Comment by hoscarcito** [comment]
I know is an old issue, but this (I guess) is giving me problems too with BackHandler on Android. It doesn't allow you to go back if you are in a screen inside a stack inside a drawer, because there's no proper route in the props in the Root where you define the helpers.

**Comment by kelset** [comment]
Hi @hoscarcito, can you open a new issue that follows the template so that it can be "pinpointed" properly?

I'd prefer to close this since it's a really long & old issue, and again it's more related to having an example than an actual issue. @julesmoretti let me know if you think otherwise.

**Comment by hoscarcito** [comment]
Yeah, no problem! But I don't think its like to have an example, because nobody know if its possible. I've found a workaround, but it takes to look for a specific path and check all the time if you are there and then close the app. But if not, you just go back. The only problem with this approach is that the "go back" takes you to the parent navigator index, not to the previous screen.
This week I'll open a new issue with all this clear. Thanks @kelset

**Comment by mcdermed** [comment]
@hoscarcito Did you open that issue?

**Comment by hoscarcito** [comment]
@mcdermed Sorry I didn't have the time. Plus, I think I made a mistake commenting here because my problem is not related with Redux.

Btw, I have it working with this configuration:

RootNav (StackNav)
- Login
- MainNav (DrawerNav)
  - Newsfeed 
  - Profile (StackNav)
  - Settings (StackNav)

After login I reset the RootNav to point to MainNav. 

Some explanation: If you are somewhere in the Profile StackNav when you click on the back button it takes you back in the stack until you reach the end of the stack, and then it takes you to the index 0 of the drawer (Newsfeed). Is that clear?

ShouldCloseApp check all the nested indexes.

Here the code:

```
import { Platform, BackHandler } from 'react-native';
import { addNavigationHelpers, NavigationActions } from 'react-navigation';

class AppNavigator extends Component {
  componentWillMount() {
    if (Platform.OS !== 'android') return;
    BackHandler.addEventListener('hardwareBackPress', this.handleBackPress);
  }

  componentWillUnmount() {
    if (Platform.OS !== 'android') return;
    BackHandler.removeEventListener('hardwareBackPress', this.handleBackPress);
  }

  // MainNav -> DrawerClose -> Profile
  //shouldCloseApp = () => this.props.nav.routes[0].routes[0].index === 0;

  shouldCloseApp = (nav) => {
    if (nav.index > 0) return false;

    if (nav.routes) {
      return nav.routes.every(this.shouldCloseApp);
    }
    return true;
  }

  goBack = () => this.props.dispatch(NavigationActions.back());

  closeDrawer = () => this.props.dispatch(NavigationActions.navigate({
    routeName: 'DrawerClose',
  }));

  handleBackPress = () => {
    if (isDrawerOpen(this.props.nav)) {
      this.closeDrawer();
      return true;
    }
    if (this.shouldCloseApp(this.props.nav)) {
      return false;
    }
    this.goBack();
    return true;
  }
```

I hope it helps!

**Comment by mcdermed** [comment]
Nice. Thanks for the detailed explanation! 👍
